### ISSUE TITLE
[RFC] Custom Importers

### ISSUE THREAD

**Issue by xzyfer** [sass/libsass#1034]
This is a RFC for the custom import C API.

This feature is currently experimental.

## Opening notes

Recently @mgreter has done some great work on the C API including, but not limited too, the multiple custom importer API https://github.com/sass/libsass/pull/1000. This RFC is by no means intended to be a criticism of this work.

## Summary

Custom importers are in charge of taking paths passed to `@import` and finding the appropriate Sass code for those paths. Importers could be added to load from a database, over HTTP, or use a different file naming scheme than what Sass expects. An `@import` can match only one importer.

## Scope

The scope of this RFC is limited to use cases for a custom importer API, and the relevant implementation details.

To that end, if you wish to follow this discussion please click the `Subscribe` button to the right than `+1`ing (these will be removed).

## Prior work

As custom importers exist in both Ruby Sass and Libsass I'll take a moment to summaries their current behaviours respectively to the best of my knowledge (please correct me on any details I have mistaken).

### Ruby Sass

Importers are resolved sequentially in the order supplied in the [:load_paths array](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#load_paths-option) (citation needed).

If an importer does handle the given path it returns the Sass content in an [Engine](http://sass-lang.com/documentation/Sass/Engine.html).
If an importer does not handle the given path, or the path is not readable, it returns `nil`.
If all importers skip the given path Ruby Sass falls back to the [Filesystem Importer](http://sass-lang.com/documentation/Sass/Importers/Filesystem.html).

An `@import` can only return only a single Engine. However since an Engine contains a Sass string, that string can be the concatenated results of many files i.e. [sass-globbing](https://github.com/chriseppstein/sass-globbing/blob/master/lib/sass/globbing/importer.rb#L83-L97)
### Libsass

Importers are resolved in a priority order (highest priority first).

If an importer does handle the given path it returns either a list of zero or imports, or an error. The returned list can contain either paths (i.e. path rewriting), or the Sass content.
If an importer does not handle the given path, or the path is not readable, it returns `0`.
If all importers skip the given path Libsass fallsback to the filesystem.

An `@import` can return one or more files.

### Resolution

The above is intentially a high level summary in order to focus on the behaviour.

The intent of this RFC is not to normalise the behvaiour of the two implementations, but instead to ensure we're

- addressing existing and future use cases
- fully leveraging existing experience from the Ruby Sass ecosystem
- not deviating from _the Sass way_ 

## Uses cases identified

Given an `@import` path we want to be able to

- resolve it to the filesystem and import it
- resolve it to the network and import it
- resolve the glob to multiple file and load them 

/cc @mgreter @am11 @chriseppstein @drewwells

**Comment by chriseppstein** [comment]
Clarification about Ruby Sass:

> If all importers skip the given path Ruby Sass falls back to the Filesystem Importer.

This is not actually true. The load path is normalized so that it is an array of importers. Strings on the load path are used to construct a filesystem importer as the root directory. The only fallback behavior is that there's a default load path array if not provided. The `:filesystem_importer` option exists to allow a user to provide a different class than the Sass default importer.

> An @import can only return only a single Engine. However since an Engine contains a Sass string, that string can be the concatenated results of many files i.e. sass-globbing

Actually, the way sass globbing works is by returning a file that is just a bunch of import statements.

The tricky thing that we have in ruby sass is the ability to tap into all instances of importers by way of ruby's class callbacks. This is how I was able to implement import-once that applies to all importers. since importers do not chain themselves in ruby. As such, it may make sense to have importers accept a continuation function that they can call when they need to delegate the lookup to the next importer.

**Comment by chriseppstein** [comment]
# Relative resolution

Sass has the semantics of importing that any import statement is first interpreted relative to the current file. This makes files very portable even across different importers. For instance you can write and test a bunch of Sass files locally but then consume them over http because the imports themselves are local to the current file (This semantic is similar to how urls on the web works).

Maybe I'm missing it, but I don't see how relative importing is working in libsass.

In ruby sass we know which importer instance was used to import each file and we ask that importer to resolve any imports relative to it first. failing that, the normal importer resolution algorithm kicks in.

**Comment by drewwells** [comment]
@chriseppstein I'm confused on the mention of the relative http importing.

It appears Ruby Sass ignores imports with `http` in them and instead uses CSS to resolve these declarations: http://sass-lang.com/documentation/file.SASS_REFERENCE.html#import 

Can you clarify what you mean by relative imports for non-files?

**Comment by mgreter** [comment]
@chriseppstein You can query the import stack which will get you the [resolved (absolute)] [2] and [relative paths] [3]. From there on it is currently left to the implementor to create the desired location from that info. But we now also expose the file [resolving algorithm] [4] via another C-API function ([not documented yet] [1]). 

```C++
sass_importer(const char* cur_path, Sass_Importer_Entry cb, struct Sass_Compiler* comp)
{
  void* cookie = sass_importer_get_cookie(cb);
  struct Sass_Import* previous = sass_compiler_get_last_import(comp);
  const char* prev_path = sass_import_get_path(previous);
}
```

The continuation is signaled by returning `null` (which is not the same as returning an empty import list). And importers are not "chained", they are just called in order until one returns not `null`.

Another use-case is the ability to pre-process imports (sass2scss could have been done as an importer for `*.sass` files). Currently libsass will not treat the initial load from a `file_context` like an import, so it is not possible to intercept this load atm. This would allow us to implement the interal importer like any other importer plugin (with the lowest priority). Would IMO be more correct from a "seperation of concerns" standpoint.

A word about the importer `priority`: Mainly needed because we can have more importers from multiple plugins. We could come up with another scheme, like IDs for plugins with some kind of dependency system. IMO this will get overly complex quite fast. I agree that consumers like node-sass or perl-libsass should rarely make use of the priority feature, so I suggest them to use a regular array for this option. Plugins should play nicely with this rule by always using a `high` priority (we could enforce this if desired), so they run before any "local" custom importers.

I guess the `priority` discussion will be continued in the "custom headers" RFC ...

[1]: https://github.com/sass/libsass/blob/master/sass_context.cpp#L814-L816
[2]: https://github.com/sass/libsass/blob/master/sass_functions.cpp#L158
[3]: https://github.com/sass/libsass/blob/master/sass_functions.cpp#L159
[4]: https://github.com/sass/libsass/blob/master/sass.h#L59

**Comment by chriseppstein** [comment]
@drewwells What I mean is that an importer can be made to import over HTTP. The file doing the importing from my local project may have an import statement like `@import "bootstrap.com/bootstrap";`

The sass file that is returned from that URL might then have an import statement like `@import "config";`. This import statement must necessarily resolve to `"bootstrap.com/config";` but only the importer that understands http semantics can know how to properly resolve that relative import.

**Comment by mgreter** [comment]
@chriseppstein I'm not 100% sure on this, but it might be possible to achieve your use-case. Importers should return the resolved/absolute location, so the next import has all the information needed to resolve any relative links inside it. When a http importer is invoked, it should see that its parent context was a http url and should create a new absolute link from the given (optional) relative import path. If any more imports occur in the returned content, the next run will get that previously created absolute link as its previous/parent path. If I understand correctly, ruby sass basically does the resolving in a seperate step and probably passes the already resolved path to the importer. In our current implementation each importer has to create the final location on its own.

This does need more testing, but that's why the whole thing is marked experimental!

**Comment by chriseppstein** [comment]
@mgreter the main difference I see is that Sass guarantees that relative resolution happens first. This is really important. Because the file on that webserver should resolve it's local file before some file "config.scss" on my local filesystem, even if my local filesystem has a higher precendence in the resolved list of importers.

**Comment by mgreter** [comment]
This is possible, since importers are queried first if they want to handle an import url. But they need to do the url mangling by themself in the same step. The http importer should have a higher priority and only accept links that have a http protocol after previous path is applied to possible relative import. Libsass only get active if all importers return `null`.

**Comment by chriseppstein** [comment]
@mgreter I feel like you're missing my point here. Did you read the part where I explained how relative resolution precedes the normal lookup logic? This isn't a matter of "giving the importer enough priority". A low priority importer takes precedence over a high one for the relative resolution phase.

**Comment by drewwells** [comment]
Each import engine is passed the current raw import string `cur_path` and previous import `Sass_Import_Entry`. Relative import would be `sass_import_get_base(previous) + cur_path`, which any importer can implement.

Global relative resolution on every import would cause unintended problems. For instance, if we're 3 directories deep and call `@import "myplugin"`, `myplugin` would get translated to `full/path/dir/myplugin` which would require complex path parsing or ignored by a high priority plugin importer expecting to be passed "myplugin". The priority system and passing raw import information to the engine avoids these problems and is a clean way of negotiating imports across multiple engines while supporting specialized operations like relative resolution.

**Comment by mgreter** [comment]
@chriseppstein I did read that part more than once and I think I understand your use-case. But I still think this is supported by libsass. We do not separate the url resolving and loading, we combine them in one step. I don't see a use-case where this should cause any issues!

The url resolving is actually a chain, since the content returned by one importer can have more imports, which will then be resolved by another (or the same) importer. The `prev_path` will always point to the absolute import location of the previous importer. So lets take the example with a file importer:

```css
echo "@import "inc.scss";" >> test.scss
echo "div { color: green; }" >> inc.scss
```

Without any custom importers involved, the file "inc.scss" get resolved relative to "test.scss", which is resolved to `cwd`. Now lets imagine we move these files to a webserver and create an local "loader":

```css
echo "@import "http://www.test.com/test.scss" >> load.scss
```

The http importer has a high priority, so it is queried first in this example. It gets `http://www...` as the `cur_path`. It now knows that it can handle that location, fetches it via http (returns the content "`@import` "inc.scss";). Now libsass will parse the returned content and sees another `@import`. It will again call all importers in order and first is the http importer. It will get "inc.scss" as `cur_path` and it can lookup the last import path, which would be `http://www.test.com/test.scss`. Now from this info we expect the http loader to know that it should load `http://www.test.com/inc.scss` (it has to handle the path mangling by itself). The mangled path is part of the expected/returned importer struct. You actually query exactly those structs via the importer stack.

So you can actually access the whole import chain, which will get you access to the actual import path (as it was mentioned in the import statement) and the resolved location (which is the location returned by the importer that "resolved" the import). We expect each importer to return a absolute path if it has resolved a import, so the next importer knows what the loading root was (it could be reconstructed from the whole stack, but we expect each importer to return the resolved absolute location).

I hope we talk about the same thing, but if not you may post some more specific use-cases!

**Comment by mgreter** [comment]
IMHO the current implementation should support the use cases discussed here pretty well. I was able to quite easily implement http loading with perl-libsass (as shown at http://libsass.ocbnet.ch/srcmap/) once I fixed a few path handling related bugs (included in 3.3.0). //CC @chriseppstein 

For completeness I will attach the perl implementation for the url importer:

```perl
sub fetch_url
{
  my ($uri) = @_;
  foreach my $tmpl ("%s", "%s.scss", "_%s.scss") {
    my $variant = new URI::URL $uri->abs;
    my @paths = $variant->path_components;
    $paths[$#paths] = sprintf($tmpl, $paths[$#paths]);
    $variant->path_components(@paths);
    my $ff = File::Fetch->new(
      uri => $variant->abs,
      tempdir_root => '/tmp'
    );
    my $where = $ff->fetch( to => \my $scalar );
    unlink $where if defined $where; # unlink tmp
    next unless (defined $scalar && $where);
    return [ [ $variant->abs, $scalar ] ];
  }
  return undef;
}

sub importer
{
  # root is the absolute path of the previous import
  my ($import, $root) = @_;
  # create a new url (root will only apply to relative imports)
  my $uri = URI::URL->new($import, $root)->abs;
  # check if we have a url scheme to handle
  if ($uri->scheme && $uri->scheme ne 'file') {
    my $result = $cache->get($uri->abs->as_string);
    unless ( defined $result ) {
      $result = fetch_url($uri) || 0;
      $cache->set($uri->abs->as_string, $result, $expiration);
    } else {
      # print STDERR "GOT CACHED ", $result, "\n";
    }
    return $result;
  }
  else {
    if (!($uri =~ m!^\Q$chrootE!)) {
      # abort if access points outside of chroot directory
      return [ [ $uri, 'IMPORT { ERR: "Access violation!"; }' ] ];
    }
  }
  # try other importers
  return undef;
}
```

**Comment by mgreter** [comment]
IMO this can be closed. Until someone points out why we should separate the path resolution and loading! We do support relative path resolutions as ruby sass does, as my tests and sample implementations should show.
